### ISSUE TITLE
Add an option to decorate code with GeneratedCodeAttribute (C#)

### ISSUE THREAD

**Issue by justmara** [grpc/grpc#24594]
### Is your feature request related to a problem? Please describe.
It's rather difficult to separate autogenerated by grpc.tool code from regular project code when using static analysis tools. Since there is no way to fix/alter code generated by grpc.tools we'd prefer to exclude that code from analysis at all instead of playing with rules/filenames combinations trying to hide particular rules/issues in these files. Also we want to exclude this code from coverage analysis, for example, because we do not mean to test grpc.tool itself and it produces pretty much code that affects whole `coverage %` picture.

### Describe the solution you'd like
There is a system [GeneratedCodeAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.codedom.compiler.generatedcodeattribute?view=netstandard-1.0) in .net, which were made exactly for such scenarios. So it would be nice to have an option to add decorate classes (or even namespaces?) generated by tool with this attribute.

### Describe alternatives you've considered
We're now using some _magic-word style_ with class/namespace special namings which is hard to maintain between projects.

### Additional context
When using [coverlet](https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/VSTestIntegration.md) for example one can simply add rule, that excludes classes by attribute and `GeneratedCodeAttribute` is used even in sample configuration.

**Comment by jtattermusch** [comment]
We currently generate 
https://github.com/grpc/grpc/blob/2dc98b1e4cfa433140bc1c7af0a99b718720331a/src/csharp/Grpc.Examples/MathGrpc.cs#L21
and 
https://github.com/grpc/grpc/blob/2dc98b1e4cfa433140bc1c7af0a99b718720331a/src/csharp/Grpc.Examples/Math.cs#L6
(actually that file is generated by protoc code from https://github.com/protocolbuffers/protobuf)
 
feel free to suggest what the diff for those files could look like (or feel free to create a PR).

**Comment by justmara** [comment]
```
    public static partial class Math {

...

    [grpc::BindServiceMethod(typeof(Math), "BindService")]
    public abstract partial class MathBase

...

    public partial class MathClient : grpc::ClientBase<MathClient>
```
class declarations would become
```
    [GeneratedCode]
    public static partial class Math {

...
    [GeneratedCode]
    [grpc::BindServiceMethod(typeof(Math), "BindService")]
    public abstract partial class MathBase

...

    [GeneratedCode]
    public partial class MathClient : grpc::ClientBase<MathClient>
```
This way static code analyzers, coverage analyzer could skip these autogenerated files with easy-setup standard instructions

**Comment by jtattermusch** [comment]
> ```
>     public static partial class Math {
> 
> ...
> 
>     [grpc::BindServiceMethod(typeof(Math), "BindService")]
>     public abstract partial class MathBase
> 
> ...
> 
>     public partial class MathClient : grpc::ClientBase<MathClient>
> ```
> 
> class declarations would become
> 
> ```
>     [GeneratedCode]
>     public static partial class Math {
> 
> ...
>     [GeneratedCode]
>     [grpc::BindServiceMethod(typeof(Math), "BindService")]
>     public abstract partial class MathBase
> 
> ...
> 
>     [GeneratedCode]
>     public partial class MathClient : grpc::ClientBase<MathClient>
> ```
> 
> This way static code analyzers, coverage analyzer could skip these autogenerated files with easy-setup standard instructions.
> optionally `using System.CodeDom.Compiler;` must be added to top-level `using`-block, or these attributes could be placed as `[System.CodeDom.Compiler.GeneratedCode]`

we'd probably want the annotation to be `[global::System.CodeDom.Compiler.GeneratedCode]`. The requirement would be that this class is available on all the target frameworks where one might use the generated protobuf code (something that needs to be checked first).

As said, this would require 1 PR to update https://github.com/protocolbuffers/protobuf/tree/master/src/google/protobuf/compiler/csharp

and another PR to update 
https://github.com/grpc/grpc/tree/master/src/compiler

I can review a PR if someone contributes it, but I likely won't have cycles to implement this myself. 

(note that we already have `[global::System.Diagnostics.DebuggerNonUserCodeAttribute]` in several places.

**Comment by StingyJack** [comment]
Linking to https://github.com/coverlet-coverage/coverlet/issues/1141   so go away stale bot.

**Comment by jtattermusch** [comment]
Isn't there a code construct that would allow to mark the entire `*Grpc.cs` file as code generated? That would make implementing this much easier.  I don't see why we would want to apply an attribute to each method, when in fact the entire file is code-generated.

**Comment by StingyJack** [comment]
@jtattermusch - not all analyzers look at the source code. The attribute will be available to both source and compiled.

**Comment by jtattermusch** [comment]
Fixed by #26164 and protocolbuffers/protobuf#8594
